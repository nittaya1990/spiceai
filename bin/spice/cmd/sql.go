/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package cmd

import (
	"log/slog"
	"os"

	"github.com/spf13/cobra"
	"github.com/spiceai/spiceai/bin/spice/pkg/context"
	"github.com/spiceai/spiceai/bin/spice/pkg/util"
)

var sqlCmd = &cobra.Command{
	Use:   "sql",
	Short: "Start an interactive SQL query session against the Spice.ai runtime",
	Example: `
$ spice sql
Welcome to the Spice.ai SQL REPL! Type 'help' for help.

show tables;  -- list available tables
sql> show tables
+---------------+--------------------+---------------+------------+
| table_catalog | table_schema       | table_name    | table_type |
+---------------+--------------------+---------------+------------+
| datafusion    | public             | tmp_view_test | VIEW       |
| datafusion    | information_schema | tables        | VIEW       |
| datafusion    | information_schema | views         | VIEW       |
| datafusion    | information_schema | columns       | VIEW       |
| datafusion    | information_schema | df_settings   | VIEW       |
+---------------+--------------------+---------------+------------+
`,
	Args: cobra.ArbitraryArgs,
	Run: func(cmd *cobra.Command, args []string) {
		rtcontext := context.NewContext()

		_, err := rtcontext.Version()
		if err != nil {
			slog.Error("Failed to run `spice sql`: The Spice runtime is not installed. Run `spice install` and retry.")
			return
		}

		spiceArgs := []string{"--repl"}

		if rootCertPath, err := cmd.Flags().GetString("tls-root-certificate-file"); err == nil && rootCertPath != "" {
			args = append(args, "--tls-root-certificate-file", rootCertPath)
		}

		if apiKey, err := cmd.Flags().GetString("api-key"); err == nil && apiKey != "" {
			args = append(args, "--api-key", apiKey)
		}

		if userAgent, err := cmd.Flags().GetString("user-agent"); err == nil && userAgent != "" {
			args = append(args, "--user-agent", userAgent)
		}

		if cacheControl, err := cmd.Flags().GetString("cache-control"); err == nil && cacheControl != "" {
			args = append(args, "--cache-control", cacheControl)
		}

		if flight, err := cmd.Flags().GetString("flight-endpoint"); err == nil && flight != "" {
			args = append(args, "--repl-flight-endpoint", flight)
		}

		if http, err := cmd.Flags().GetString("http-endpoint"); err == nil && http != "" {
			args = append(args, "--http-endpoint", http)
		}

		args = append(spiceArgs, args...)

		execCmd, err := rtcontext.GetRunCmd(args)
		if err != nil {
			slog.Error("getting run command", "error", err)
			os.Exit(1)
		}

		execCmd.Stderr = os.Stderr
		execCmd.Stdout = os.Stdout
		execCmd.Stdin = os.Stdin

		err = util.RunCommand(execCmd)
		if err != nil {
			slog.Error("running command", "error", err, "command", execCmd.String())
			os.Exit(1)
		}
	},
}

func init() {
	sqlCmd.Flags().String("tls-root-certificate-file", "", "The path to the root certificate file used to verify the Spice.ai runtime server certificate")
	sqlCmd.Flags().String("user-agent", "", "The user agent to use for all requests")
	sqlCmd.Flags().String("cache-control", "cache", "Control whether the results cache is used for queries. [possible values: cache, no-cache]")
	sqlCmd.Flags().String("flight-endpoint", "", "Specifies the runtime Flight endpoint. Defaults to http://localhost:50051")
	sqlCmd.Flags().String("http-endpoint", "", "Specifies the runtime HTTP endpoint. Defaults to http://localhost:8090")

	RootCmd.AddCommand(sqlCmd)
}
