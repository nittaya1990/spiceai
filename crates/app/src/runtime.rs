/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::{fmt::Display, str::FromStr, sync::Arc};

use spicepod::component::runtime::UserAgentCollection;

use super::App;

impl App {
    /// Get a parameter from the app's runtime params, with a default value if the parameter is not set or is not valid.
    ///
    /// Returns `default_value` if the parameter is not set or is not valid.
    ///
    /// If the parameter is set but is not valid, logs a warning and returns `default_value`.
    #[must_use]
    pub fn get_runtime_param<T>(app: &Option<Arc<Self>>, param: &str, default_value: T) -> T
    where
        T: Display + FromStr,
    {
        let Some(value) = app.as_ref().and_then(|app| app.runtime.params.get(param)) else {
            return default_value;
        };

        if let Ok(parsed_value) = value.parse::<T>() {
            parsed_value
        } else {
            eprintln!("runtime.params.{param} is not valid, defaulting to {default_value}");
            default_value
        }
    }

    #[must_use]
    pub fn user_agent_collection(&self) -> UserAgentCollection {
        self.runtime.telemetry.user_agent_collection
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::AppBuilder;
    use std::{collections::HashMap, sync::Arc};

    fn create_app_with_params(params: HashMap<String, String>) -> Arc<App> {
        Arc::new(AppBuilder::new("test").with_runtime_params(params).build())
    }

    #[test]
    fn test_get_runtime_param() {
        // Test case 1: Parameter is not set
        let app = Some(create_app_with_params(HashMap::new()));
        assert!(App::get_runtime_param(&app, "test_param", true));
        assert!(!App::get_runtime_param(&app, "test_param", false));

        // Test case 2: Parameter is set to "true"
        let mut params = HashMap::new();
        params.insert("test_param".to_string(), "true".to_string());
        let app = Some(create_app_with_params(params));
        assert!(App::get_runtime_param(&app, "test_param", false));

        // Test case 3: Parameter is set to "false"
        let mut params = HashMap::new();
        params.insert("test_param".to_string(), "false".to_string());
        let app = Some(create_app_with_params(params));
        assert!(!App::get_runtime_param(&app, "test_param", true));

        // Test case 4: Parameter is set to an invalid boolean value
        let mut params = HashMap::new();
        params.insert("test_param".to_string(), "not_a_bool".to_string());
        let app = Some(create_app_with_params(params));
        assert!(App::get_runtime_param(&app, "test_param", true));
        assert!(!App::get_runtime_param(&app, "test_param", false));

        // Test case 5: App is None
        assert!(App::get_runtime_param(&None, "test_param", true));
        assert!(!App::get_runtime_param(&None, "test_param", false));
    }
}
