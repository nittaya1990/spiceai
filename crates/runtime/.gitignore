abc-sqlite.db-wal
abc-sqlite.db-shm
