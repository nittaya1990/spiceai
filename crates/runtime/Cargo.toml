[package]
description = "The spiced runtime"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "runtime"
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[dependencies]
app = { path = "../app" }
arrow.workspace = true
arrow-csv = "53.3.0"
arrow-flight = { workspace = true, features = ["flight-sql-experimental"] }
arrow-ipc.workspace = true
arrow-json.workspace = true
arrow-schema = { workspace = true, features = ["serde"] }
arrow_sql_gen = { path = "../arrow_sql_gen" }
arrow_tools = { path = "../arrow_tools" }
async-openai.workspace = true
async-stream.workspace = true
async-trait.workspace = true
aws-config = { version = "1.1.10", optional = true }
aws-sdk-dynamodb = { version = "1", optional = true }
aws-sdk-secretsmanager = { version = "1", optional = true }
aws-sdk-sts = { version = "1", optional = true }
axum.workspace = true
axum-extra = { version = "0.9.6", features = ["typed-header"] }
base64.workspace = true
bytes.workspace = true
cache = { path = "../cache" }
chrono = { version = "0.4.38" }
clap.workspace = true
clickhouse-rs = { workspace = true, optional = true }
csv = "1.3.0"
dashmap = "6.1.0"
data_components = { path = "../data_components" }
datafusion.workspace = true
datafusion-federation = { workspace = true }
datafusion-federation-sql = { workspace = true }
datafusion-functions-json = { workspace = true }
datafusion-table-providers = { workspace = true }
db_connection_pool = { path = "../db_connection_pool" }
document_parse = { path = "../document_parse" }
dotenvy.workspace = true
duckdb = { workspace = true, features = [
    "bundled",
    "r2d2",
    "vtab",
    "vtab-arrow",
], optional = true }
flight_client = { path = "../flight_client" }
fundu = { workspace = true }
futures.workspace = true
globset.workspace = true
governor = "0.8.0"
graph-rs-sdk = { workspace = true, optional = true }
graphql-parser.workspace = true
headers-accept = "0.1.3"
http.workspace = true
http-body-util = "0.1.2"
hyper = "1.6.0"
hyper-util = { version = "0.1.6", features = ["service"] }
iceberg.workspace = true
iceberg-catalog-rest.workspace = true
imap = { workspace = true, optional = true }
indexmap.workspace = true
itertools.workspace = true
jsonwebtoken = "9.3.0"
keyring = { version = "3.6.1", features = [
    "apple-native",
    "windows-native",
    "linux-native",
], optional = true }
llms = { path = "../llms" }
logos = "0.15.0"
mcp-client = { git = "https://github.com/block/goose", package = "mcp-client", optional = true }
mcp-core = { git = "https://github.com/block/goose", package = "mcp-core", optional = true }
mediatype = "0.19.18"
model_components = { path = "../model_components" }
mysql_async = { workspace = true, optional = true }
notify = "7.0.0"
ns_lookup = { path = "../ns_lookup" }
object_store = { workspace = true, features = ["aws", "http", "azure"] }
once_cell = "1.19.0"
opentelemetry.workspace = true
opentelemetry-prometheus.workspace = true
opentelemetry-proto = { version = "0.27", features = [
    "gen-tonic-messages",
    "gen-tonic",
    "metrics",
] }
opentelemetry_sdk.workspace = true
otel-arrow = { path = "../otel-arrow" }
package = { path = "../package" }
percent-encoding.workspace = true
pin-project.workspace = true
prometheus.workspace = true
prometheus-parse = "0.2.5"
prost = { version = "0.13.1", default-features = false, features = [
    "prost-derive",
] }
rand = "0.8.5"
regex.workspace = true
reqwest.workspace = true
runtime-auth = { path = "../runtime-auth" }
rusqlite = { workspace = true, optional = true }
rustls.workspace = true
rustls-pemfile.workspace = true
schemars = "0.8.19"
secrecy.workspace = true
serde = { workspace = true, features = ["derive"] }
serde_json.workspace = true
snafu.workspace = true
snowflake-api = { workspace = true, optional = true }
spicepod = { path = "../spicepod" }
ssh2 = { workspace = true, optional = true }
suppaftp = { workspace = true, optional = true }
telemetry = { path = "../telemetry" }
tiberius = { workspace = true, optional = true }
tokio.workspace = true
tokio-rusqlite = { workspace = true, optional = true }
tokio-rustls = "0.26.0"
tokio-stream.workspace = true
tokio-util = { workspace = true, optional = true }
tonic.workspace = true
tonic-health.workspace = true
tower.workspace = true
tower-http.workspace = true
tracing.workspace = true
tracing-futures.workspace = true
tract-core = "0.21.0"
url = "2.5.4"
util = { path = "../util" }
utoipa = { version = "5", features = ["axum_extras"], optional = true }
utoipa-swagger-ui = { version = "8", features = ["axum"], optional = true }

uuid = { workspace = true, features = ["v7"] }
x509-certificate.workspace = true

[dev-dependencies]
anyhow = "1.0.86"
async-graphql = "7.0.5"
async-graphql-axum = "=7.0.13"
azure_core = "0.21.0"
azure_storage = "0.21.0"
azure_storage_blobs = "0.21.0"
bollard = "0.18.1"
dirs = "5.0.1"
flightrepl = { path = "../flightrepl" }
hf-hub = { workspace = true }
insta.workspace = true
jsonpath-rust = "0.7.3"
opentelemetry_sdk = { workspace = true, features = ["rt-tokio-current-thread"] }
paste.workspace = true
scopeguard = "1.2.0"
serde_json = { workspace = true, features = ["preserve_order"] }
spice-cloud = { path = "../spice_cloud" }
test-framework = { path = "../test-framework" }
tokio = { workspace = true, features = ["time", "test-util"] }
tracing-opentelemetry.workspace = true
tracing-subscriber.workspace = true

[features]
aws-secrets-manager = [
    "dep:aws-config",
    "dep:aws-sdk-secretsmanager",
    "dep:aws-sdk-sts",
]
clickhouse = [
    "dep:clickhouse-rs",
    "db_connection_pool/clickhouse",
    "arrow_sql_gen/clickhouse",
    "data_components/clickhouse",
]
cuda = ["llms/cuda"]
databricks = ["data_components/databricks"]
debezium = ["data_components/debezium"]
default = ["keyring-secret-store", "aws-secrets-manager", "sharepoint"]
delta_lake = ["data_components/delta_lake"]
dev = ["openapi", "otel-arrow/dev", "mcp"]
dremio = []
duckdb = ["dep:duckdb", "db_connection_pool/duckdb", "data_components/duckdb"]
dynamodb = ["dep:aws-sdk-dynamodb", "data_components/dynamodb"]
flightsql = ["data_components/flightsql"]
ftp = ["dep:suppaftp", "dep:ssh2"]
imap = ["data_components/imap", "dep:imap"]
keyring-secret-store = ["dep:keyring"]
mcp = ["dep:mcp-client", "dep:mcp-core"]
metal = ["llms/metal"]
models = ["model_components/full"]
mssql = ["dep:tiberius", "dep:tokio-util", "data_components/mssql"]
mysql = ["dep:mysql_async", "db_connection_pool/mysql", "data_components/mysql"]
odbc = ["db_connection_pool/odbc", "data_components/odbc"]
openapi = ["dep:utoipa-swagger-ui", "dep:utoipa"]
postgres = [
    "db_connection_pool/postgres",
    "data_components/postgres",
    "datafusion-table-providers/postgres",
]
postgres-write = []
sharepoint = ["data_components/sharepoint", "dep:graph-rs-sdk"]
snowflake = [
    "dep:snowflake-api",
    "db_connection_pool/snowflake",
    "data_components/snowflake",
]
spark = ["data_components/spark_connect"]
sqlite = [
    "dep:tokio-rusqlite",
    "db_connection_pool/sqlite",
    "data_components/sqlite",
    "dep:rusqlite",
]

[[bench]]
harness = false
name = "bench"

[[bench]]
harness = false
name = "vector_search"

[target.'cfg(windows)'.dependencies]
winver = "1.0.0"
