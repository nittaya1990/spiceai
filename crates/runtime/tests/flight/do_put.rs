/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::{
    net::{IpAddr, Ipv4Addr, SocketAddr},
    num::NonZeroU32,
    pin::Pin,
    sync::Arc,
    task::{Context, Poll},
    time::Duration,
};

use crate::{
    init_tracing,
    utils::{test_request_context, wait_until_true},
};
use arrow::array::{Int32Array, RecordBatch, StringArray};
use arrow_flight::{
    encode::FlightDataEncoderBuilder, error::FlightError, FlightClient, FlightDescriptor, PutResult,
};
use arrow_schema::{DataType, Field, Schema, SchemaRef};
use datafusion::sql::TableReference;
use futures::{
    stream::{self, TryStreamExt},
    Stream,
};
use governor::Quota;
use rand::Rng;
use runtime::{
    accelerated_table::refresh::Refresh, auth::EndpointAuth,
    component::dataset::acceleration::Acceleration, config::Config, datafusion::DataFusion,
    flight::RateLimits, internal_table::create_internal_accelerated_table, secrets::Secrets,
    Runtime,
};
use runtime_auth::{api_key::ApiKeyAuth, FlightBasicAuth};
use spicepod::component::runtime::ApiKey;
use tokio::{
    sync::RwLock,
    time::{sleep, timeout},
};
use tokio_stream::StreamExt;
use tonic::transport::Channel;

const LOCALHOST: IpAddr = IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1));

#[tokio::test]
async fn test_flight_do_put_basic() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(Some("integration=debug,info"));

    test_request_context()
        .scope(async {
            let auth = Arc::new(ApiKeyAuth::new(vec![ApiKey::parse_str("valid:rw")]))
                as Arc<dyn FlightBasicAuth + Send + Sync>;

            let (channel, df) = start_spice_test_app(Some(auth), None).await?;

            let mut client = create_flight_client(channel, Some("valid"))?;

            let test_record_batch = test_record_batch()?;

            let response = write_record_batches(
                &mut client,
                // simulate two record batches / two FlightData messages
                vec![test_record_batch.clone(), test_record_batch].into_iter(),
            )
            .await?;

            let response_str = format!("{response:?}");
            insta::assert_snapshot!("do_put_basic_reponse", response_str);

            let query = df
                .query_builder("SELECT * from my_table")
                .build()
                .run()
                .await?;

            let results: Vec<RecordBatch> = query.data.try_collect::<Vec<RecordBatch>>().await?;
            let results_str =
                arrow::util::pretty::pretty_format_batches(&results).expect("pretty batches");
            insta::assert_snapshot!("do_put_basic_table_content", results_str);

            Ok(())
        })
        .await
}

#[tokio::test]
async fn test_do_put_stream_error() -> Result<(), Box<dyn std::error::Error>> {
    let auth = Arc::new(ApiKeyAuth::new(vec![ApiKey::parse_str("valid:rw")]))
        as Arc<dyn FlightBasicAuth + Send + Sync>;

    let (channel, df) = start_spice_test_app(Some(auth), None).await?;

    let mut client = create_flight_client(channel, Some("valid"))?;

    let test_record_batch = test_record_batch()?;

    let repeating_stream = RepeatingStream {
        batch: test_record_batch.clone(),
    };

    // simulate a sending the same record batch every 250ms
    let delayed_stream = repeating_stream.then(|batch| async move {
        sleep(Duration::from_millis(250)).await;
        batch
    });

    let flight_descriptor = FlightDescriptor::new_path(vec!["my_table".to_string()]);
    let flight_data_stream = FlightDataEncoderBuilder::new()
        .with_flight_descriptor(Some(flight_descriptor))
        .build(delayed_stream);

    // simulate unexpected stream termination after 3 seconds
    let result = timeout(Duration::from_secs(3), async {
        let result: Result<Vec<PutResult>, FlightError> = client
            .do_put(flight_data_stream)
            .await
            .expect("to get result stream")
            .try_collect()
            .await;
        result
    })
    .await;

    assert!(
        result.is_err(),
        "Expected an error but got a successful result"
    );

    // Verify that no data was written to the table
    let query = df
        .query_builder("SELECT * from my_table")
        .build()
        .run()
        .await?;

    let results: Vec<RecordBatch> = query.data.try_collect::<Vec<RecordBatch>>().await?;
    let results_str = arrow::util::pretty::pretty_format_batches(&results).expect("pretty batches");
    insta::assert_snapshot!("stream_error_table_content", results_str);

    Ok(())
}

#[tokio::test]
async fn test_flight_do_put_no_auth() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(Some("integration=debug,info"));

    test_request_context()
        .scope(async {
            let (channel, _df) = start_spice_test_app(None, None).await?;

            let mut client = create_flight_client(channel, None)?;

            let response =
                write_record_batches(&mut client, vec![test_record_batch()?].into_iter()).await;

            assert!(
                response.is_err(),
                "Expected an error but got a successful result"
            );

            Ok(())
        })
        .await
}

#[tokio::test]
async fn test_flight_do_put_ro_key() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(Some("integration=debug,info"));

    test_request_context()
        .scope(async {
            let auth = Arc::new(ApiKeyAuth::new(vec![ApiKey::parse_str("valid")]))
                as Arc<dyn FlightBasicAuth + Send + Sync>;

            let (channel, _df) = start_spice_test_app(Some(auth), None).await?;

            let mut client = create_flight_client(channel, Some("valid"))?;

            let response =
                write_record_batches(&mut client, vec![test_record_batch()?].into_iter()).await;

            assert!(
                response.is_err(),
                "Expected an error but got a successful result"
            );

            Ok(())
        })
        .await
}

#[tokio::test]
async fn test_flight_do_put_rate_limit() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(Some("integration=debug,info"));

    test_request_context()
        .scope(async {
            let auth = Arc::new(ApiKeyAuth::new(vec![ApiKey::parse_str("valid:rw")]))
                as Arc<dyn FlightBasicAuth + Send + Sync>;

            let write_quota = Quota::with_period(Duration::from_secs(10))
                .expect("to create quota")
                .allow_burst(NonZeroU32::new(5).expect("should convert 5 to NonZeroU32"));

            let (channel, df) = start_spice_test_app(
                Some(auth),
                Some(RateLimits::new().with_flight_write_limit(write_quota)),
            )
            .await?;

            let mut client = create_flight_client(channel, Some("valid"))?;

            let test_record_batch = test_record_batch()?;

            // simulate 5 requests to reach rate limit
            for _ in 1..=5 {
                let _ =
                    write_record_batches(&mut client, vec![test_record_batch.clone()].into_iter())
                        .await?;
            }

            // rate limit error is expected next
            assert!(
                write_record_batches(&mut client, vec![test_record_batch.clone()].into_iter())
                    .await
                    .is_err(),
                "Expected an error but got a successful result"
            );

            // wait for the rate limit reset and perform another request attempt
            sleep(Duration::from_secs(10)).await;

            let _ = write_record_batches(&mut client, vec![test_record_batch.clone()].into_iter())
                .await?;

            let query = df
                .query_builder("SELECT * from my_table")
                .build()
                .run()
                .await?;

            let results: Vec<RecordBatch> = query.data.try_collect::<Vec<RecordBatch>>().await?;
            let results_str =
                arrow::util::pretty::pretty_format_batches(&results).expect("pretty batches");
            insta::assert_snapshot!("rate_limit_table_content", results_str);

            Ok(())
        })
        .await
}

#[tokio::test]
async fn test_flight_do_put_max_rows_allowed() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(Some("integration=debug,info"));

    test_request_context()
        .scope(async {
            let auth = Arc::new(ApiKeyAuth::new(vec![ApiKey::parse_str("valid:rw")]))
                as Arc<dyn FlightBasicAuth + Send + Sync>;

            let (channel, df) = start_spice_test_app(Some(auth), None).await?;

            let mut client = create_flight_client(channel, Some("valid"))?;

            assert!(
                // Simulate a normal batch, followed by a batch that exceeds the allowed number of rows, and then another normal batch.
                write_record_batches(
                    &mut client,
                    vec![
                        test_record_batch()?,
                        large_test_record_batch()?,
                        test_record_batch()?
                    ]
                    .into_iter()
                )
                .await
                .is_err(),
                "Expected an error but got a successful result"
            );

            let query = df
                .query_builder("SELECT * from my_table")
                .build()
                .run()
                .await?;

            let results: Vec<RecordBatch> = query.data.try_collect::<Vec<RecordBatch>>().await?;
            let results_str =
                arrow::util::pretty::pretty_format_batches(&results).expect("pretty batches");
            insta::assert_snapshot!("max_rows_allowed_table_content", results_str);

            Ok(())
        })
        .await
}

#[tokio::test]
async fn test_do_put_read_timeout() -> Result<(), Box<dyn std::error::Error>> {
    let auth = Arc::new(ApiKeyAuth::new(vec![ApiKey::parse_str("valid:rw")]))
        as Arc<dyn FlightBasicAuth + Send + Sync>;

    let (channel, df) = start_spice_test_app(Some(auth), None).await?;

    let mut client = create_flight_client(channel, Some("valid"))?;

    let record_batch_1 = test_record_batch()?;
    let record_batch_2 = record_batch_1.clone();
    let record_batch_3 = record_batch_1.clone();

    let first_batch =
        stream::once(async { Ok(record_batch_1) as Result<RecordBatch, FlightError> });
    // batch with 40s delay
    let second_batch = stream::once(async {
        sleep(Duration::from_secs(40)).await;
        Ok(record_batch_2) as Result<RecordBatch, FlightError>
    });
    let third_batch =
        stream::once(async { Ok(record_batch_3) as Result<RecordBatch, FlightError> });

    let flight_descriptor = FlightDescriptor::new_path(vec!["my_table".to_string()]);
    let flight_data_stream = FlightDataEncoderBuilder::new()
        .with_flight_descriptor(Some(flight_descriptor))
        .build(first_batch.chain(second_batch).chain(third_batch));

    let result: Result<Vec<PutResult>, FlightError> = client
        .do_put(flight_data_stream)
        .await
        .expect("to get result stream")
        .try_collect()
        .await;

    assert!(
        result.is_err(),
        "Expected an error but got a successful result"
    );

    // Verify that no data was written to the table
    let query = df
        .query_builder("SELECT * from my_table")
        .build()
        .run()
        .await?;

    let results: Vec<RecordBatch> = query.data.try_collect::<Vec<RecordBatch>>().await?;
    let results_str = arrow::util::pretty::pretty_format_batches(&results).expect("pretty batches");
    insta::assert_snapshot!("read_timeout_table_content", results_str);

    Ok(())
}

async fn write_record_batches(
    client: &mut FlightClient,
    batches: impl Iterator<Item = RecordBatch>,
) -> Result<Vec<PutResult>, FlightError> {
    let flight_descriptor = FlightDescriptor::new_path(vec!["my_table".to_string()]);
    let flight_data_stream = FlightDataEncoderBuilder::new()
        .with_flight_descriptor(Some(flight_descriptor))
        .build(futures::stream::iter(batches.map(Ok).collect::<Vec<_>>()));

    let response: Vec<PutResult> = client
        .do_put(flight_data_stream)
        .await?
        .try_collect()
        .await?;

    Ok(response)
}

fn create_flight_client(
    channel: Channel,
    api_key: Option<&str>,
) -> Result<FlightClient, anyhow::Error> {
    let mut client = FlightClient::new(channel);

    if let Some(api_key) = api_key {
        client
            .add_header("authorization", &format!("Bearer {api_key}"))
            .map_err(anyhow::Error::from)?;
    };

    Ok(client)
}

async fn start_spice_test_app(
    flight_auth: Option<Arc<dyn FlightBasicAuth + Send + Sync>>,
    rate_limits: Option<RateLimits>,
) -> Result<(Channel, Arc<DataFusion>), anyhow::Error> {
    let mut rng = rand::thread_rng();
    let http_port: u16 = rng.gen_range(50000..60000);
    let flight_port: u16 = http_port + 1;
    let otel_port: u16 = http_port + 2;
    let metrics_port: u16 = http_port + 3;

    tracing::debug!(
        "Ports: http: {http_port}, flight: {flight_port}, otel: {otel_port}, metrics: {metrics_port}"
    );

    let api_config = Config::new()
        .with_http_bind_address(SocketAddr::new(LOCALHOST, http_port))
        .with_flight_bind_address(SocketAddr::new(LOCALHOST, flight_port))
        .with_open_telemetry_bind_address(SocketAddr::new(LOCALHOST, otel_port));

    let registry = prometheus::Registry::new();

    let mut rt_builder =
        Runtime::builder().with_metrics_server(SocketAddr::new(LOCALHOST, metrics_port), registry);

    if let Some(rate_limits) = rate_limits {
        rt_builder = rt_builder.with_rate_limits(rate_limits);
    }

    let rt = rt_builder.build().await;

    let df = rt.datafusion();

    let test_record_batch = test_record_batch()?;

    register_test_table(
        &df,
        test_record_batch.schema(),
        TableReference::parse_str("public.my_table"),
    )
    .await?;

    let mut auth = EndpointAuth::default();

    if let Some(flight_auth) = flight_auth {
        auth = auth.with_flight_basic_auth(flight_auth);
    }

    // Start the servers
    tokio::spawn(async move { Box::pin(Arc::new(rt).start_servers(api_config, None, auth)).await });

    // Wait for the servers to start
    tracing::info!("Waiting for servers to start...");
    wait_until_true(Duration::from_secs(10), || async {
        reqwest::get(format!("http://localhost:{http_port}/health"))
            .await
            .is_ok()
    })
    .await;

    let channel = Channel::from_shared(format!("http://localhost:{flight_port}"))?
        .connect()
        .await
        .map_err(anyhow::Error::from)?;

    Ok((channel, df))
}

fn test_record_batch() -> Result<RecordBatch, anyhow::Error> {
    let schema = Arc::new(Schema::new(vec![
        Field::new("a", DataType::Int32, false),
        Field::new("b", DataType::Utf8, false),
    ]));

    RecordBatch::try_new(
        Arc::clone(&schema),
        vec![
            Arc::new(Int32Array::from(vec![1, 2, 3])),
            Arc::new(StringArray::from(vec!["a", "b", "c"])),
        ],
    )
    .map_err(anyhow::Error::from)
}

fn large_test_record_batch() -> Result<RecordBatch, anyhow::Error> {
    let schema = Arc::new(Schema::new(vec![
        Field::new("a", DataType::Int32, false),
        Field::new("b", DataType::Utf8, false),
    ]));

    // Generate 35,000 rows of data
    let int_column = (1..=35_000).collect::<Vec<i32>>();
    let string_column = (1..=35_000)
        .map(|i| format!("row_{i}"))
        .collect::<Vec<String>>();

    RecordBatch::try_new(
        Arc::clone(&schema),
        vec![
            Arc::new(Int32Array::from(int_column)),
            Arc::new(StringArray::from(string_column)),
        ],
    )
    .map_err(anyhow::Error::from)
}

async fn register_test_table(
    datafusion: &Arc<DataFusion>,
    schema: SchemaRef,
    table_name: TableReference,
) -> Result<(), anyhow::Error> {
    let table = create_internal_accelerated_table(
        datafusion.runtime_status(),
        table_name.clone(),
        schema,
        None,
        Acceleration::default(),
        Refresh::default(),
        None,
        Arc::new(RwLock::new(Secrets::default())),
    )
    .await
    .map_err(anyhow::Error::from)?;

    datafusion
        .register_table_as_writable_and_with_schema(table_name, table)
        .map_err(anyhow::Error::from)?;

    Ok(())
}

struct RepeatingStream {
    batch: RecordBatch,
}

impl Stream for RepeatingStream {
    type Item = Result<RecordBatch, FlightError>;

    fn poll_next(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        Poll::Ready(Some(Ok(self.batch.clone())))
    }
}
