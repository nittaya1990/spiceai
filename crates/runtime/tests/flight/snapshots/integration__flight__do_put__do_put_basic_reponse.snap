---
source: crates/runtime/tests/flight/do_put.rs
expression: response_str
---
[PutResult { app_metadata: b"" }, PutResult { app_metadata: b"" }]
